packages/web/src/apollo/suspense.tsx

```diff
@@ -0,0 +1,312 @@
+  //
+  // See https://www.apollographql.com/docs/react/api/link/introduction.
+  const { getToken, type: authProviderType } = useAuth()
+
+  // `updateDataApolloLink` keeps track of the most recent req/res data so they can be passed to
+  // any errors passed up to an error boundary.
+  const data = {
+    mostRecentRequest: undefined,
+    mostRecentResponse: undefined,
+  } as any
+
+  const updateDataApolloLink = new ApolloLink((operation, forward) => {
+    const { operationName, query, variables } = operation
+
+    data.mostRecentRequest = {}
+    data.mostRecentRequest.operationName = operationName
+    data.mostRecentRequest.operationKind = query?.kind.toString()
+    data.mostRecentRequest.variables = variables
+    data.mostRecentRequest.query = query && print(operation.query)
+
+    return forward(operation).map((result) => {
+      data.mostRecentResponse = result
+
+      return result
+    })
+  })
+
+  const withToken = setContext(async () => {
+    const token = await getToken()
+
+    return { token }
+  })
+
+  const { headers, uri } = useFetchConfig()
+
+  const getGraphqlUrl = () => {
+    // @NOTE: This comes from packages/vite/src/streaming/registerGlobals.ts
+    // this needs to be an absolute url, as relative urls cannot be used in SSR
+    // @TODO (STREAMING): Should this be a new config value in Redwood.toml?
+    // How do we know what the absolute url is in production?
+    // Possible solution: https://www.apollographql.com/docs/react/api/link/apollo-link-schema/
+
+    return typeof window === 'undefined'
+      ? RWJS_ENV.RWJS_EXP_SSR_GRAPHQL_ENDPOINT
+      : uri
+  }
+
+  const authMiddleware = new ApolloLink((operation, forward) => {
+    const { token } = operation.getContext()
+
+    // Only add auth headers when there's a token. `token` is `null` when `!isAuthenticated`.
+    const authHeaders = token
+      ? {
+          'auth-provider': authProviderType,
+          authorization: `Bearer ${token}`,
+        }
+      : {}
+
+    operation.setContext(() => ({
+      headers: {
+        ...operation.getContext().headers,
+        ...headers,
+        // Duped auth headers, because we may remove the `FetchConfigProvider` at a later date.
+        ...authHeaders,
+      },
+    }))
+
+    return forward(operation)
+  })
+
+  const { httpLinkConfig, link: redwoodApolloLink, ...rest } = config ?? {}
+
+  // A terminating link. Apollo Client uses this to send GraphQL operations to a server over HTTP.
+  // See https://www.apollographql.com/docs/react/api/link/introduction/#the-terminating-link.
+  const httpLink = new HttpLink({ uri, ...httpLinkConfig })
+
+  // The order here is important. The last link *must* be a terminating link like HttpLink.
+  const redwoodApolloLinks: RedwoodApolloLinks = [
+    { name: 'withToken', link: withToken },
+    { name: 'authMiddleware', link: authMiddleware },
+    { name: 'updateDataApolloLink', link: updateDataApolloLink },
+    { name: 'httpLink', link: httpLink },
```

Comment:
FYI as part of PR to support GraphQL subscriptions over SSE (https://github.com/redwoodjs/redwood/pull/9009) there is now a new terminating link that is directional -- that is, if the operation is a Subscription then the client uses an SSELink and otherwise a classic HttpLink.

In addition, my current Fragments rework enables a Fragment registry on the Apollo client cache.

My question is -- do we refactor the shared portion of the the Suspense and non Suspense clients?

Reply:
Also, have a look are "directional composition" https://www.apollographql.com/docs/react/api/link/introduction/#directional-composition -- you can set a terminating link based on a condition ... for example if the operation is a Subscription use SSELink as the terminating link and if not HTTPLink.

You could perhaps setup a http link for streaming and swap out based on a config.

Reply:
You mean for this PR? Or something for later?

Reply:
I looked into this - we don't need to change the link based on the request. We just need to change it if its running on the server or not - so I think directional composition may not be required here.